### DISCUSSION TITLE
ChildNodesLayout position of FIRST_CHILD_BY_PARENT not respected

### DISCUSSION THREAD

**Discussion by macmarrum** [freeplane/freeplane#1074] [General]
v1.11.1-pre26

Node `1` is expected at the same Y offset from `root` as node `2`. Instead it is pushed up by its sibling.

![bug-v11-ChildNodesLayout-position-not-respected-FIRST](https://user-images.githubusercontent.com/7197131/229362199-4395c3fe-2731-4ad7-bc5d-90b2c1e5bdff.png)

[bug-v11-ChildNodesLayout-position-not-respected.mm.zip](https://github.com/freeplane/freeplane/files/11132585/bug-v11-ChildNodesLayout-position-not-respected.mm.zip)

**Comment by macmarrum** [comment]
Similarly in case of LEFT_TO_RIGHT

![bug-v11-ChildNodesLayout-position-not-respected-FIRST-LtoR](https://user-images.githubusercontent.com/7197131/229362622-b1468cae-6c15-474e-8ea6-22fab82fa485.png)

**Comment by dpolivaev** [comment]
Actually it is not a bug.
The map is laid out so that

-  if nodes 1, 2, 3 and 4 are folded, the alignment holds
-  if the user unfolds any node, it position relatively to its parent node does not change.

Therefore unfolding node 3 shifts node 1 up. It makes everything look smooth. Otherwise user would see unpleasant jumps.

**Comment by dpolivaev** [comment]
I have converted the issue to discussion so that the logic can be discussed by the community

**Comment by macmarrum** [comment]
It works as expected in v1.10

![bug-v11-ChildNodesLayout-position-not-respected-v10](https://user-images.githubusercontent.com/7197131/229370928-56e074d4-89f5-4c3f-863c-cee4587ce129.png)

[bug-v11-ChildNodesLayout-position-not-respected-v10.mm.zip](https://github.com/freeplane/freeplane/files/11132970/bug-v11-ChildNodesLayout-position-not-respected-v10.mm.zip)

**Reply by macmarrum** [reply]
Migrating from v1.10 to v1.11 will change how existing mindmaps look like. I believe it shouldn't.

And since I use FIRST / LAST a lot, if this new way is kept, I will most probably stay with v1.10.

**Reply by dpolivaev** [reply]
Unless it is a bug fix

**Reply by dpolivaev** [reply]
I can introduce an option if you find a way to make it discoverable and transparent to users

**Reply by dpolivaev** [reply]
Also if you want to convince me that you need any feature you can share more information about your use cases

**Reply by dpolivaev** [reply]
Changing the logic is just one line in the code and I have already tried it.

But you need to explain why you need to reverse FIRST and LAST alignments for sibling nodes as in your example.

**Reply by macmarrum** [reply]
Please see the post below: https://github.com/freeplane/freeplane/discussions/1074#discussioncomment-5503328

**Reply by dpolivaev** [reply]
An argument like "I will most probably stay with v1.10." does not explain your needs at all. It just makes a claim about it which I have to believe but which I have no chance to understand.

**Comment by macmarrum** [comment]
It works as expected in v1.11 when the children of **node 3** cause the branch to grow towards the bottom.

![bug-v11-ChildNodesLayout-position-not-respected-FIRST+AFTER](https://user-images.githubusercontent.com/7197131/229371547-696bd5bc-988b-49df-8775-4ab5bf3e357d.png)

Then, once **node 3** ChildNodesLayout is changed to e.g. LAST_CHILD_BY_PARENT or BEFORE_PARENT, **node 1** is forced off its original position. In my opinion, this breaks the promise of FIRST_CHILD_BY_PARENT made to **node 1**.

To me the promise is to keep **node 1** as the first one by its parent, not to be at the same coordinates regardless of whether **node 3** is folded or not or has a towards-bottom-growing branch.

I fail to see any benefit of this new behavior.

**Reply by dpolivaev** [reply]
The promise to keep the same relative coordinates after folding  / unfolding was one of the key FreeMind/Freeplane feature ever. It was broken in Version 1.10 and is fixed now.

If you use the same alignment for all nodes in a subtree, both constraints can be fulfilled.

If you want to change the kind of alignment please explain why.

**Reply by dpolivaev** [reply]
I have explicitly converted your issue to a discussion to get good arguments.
If nobody else is interested and no arguments can be given, there is no reason to make any change.

**Reply by dpolivaev** [reply]
I do not use any of these layout-related features myself. Therefore I can not have an opinion unless people who have preferences can explain their use cases. In doubt with examples.

**Comment by sifran-github** [comment]
I will try to rephrase what I understand from this thread, maybe it will help others.

If i summarize, this is a problem where two nodes grow their children in opposing directions, "on top of each other", so the question is: "which node should move and give space for the child nodes? node 1 or node 3?" This is a specific case of this problem.

## ARGUMENTS FOR: NODE 3 SHOULD MOVE DOWN
part of @macmarrum point is:
> this breaks the promise of FIRST_CHILD_BY_PARENT made to node 1

He means that the "root" node is having this property, but in fact the first child is not by his parent (root). I think that this is correct, and i do understand his conclusion that "in this case, node 3 should move down".

## ARGUMENTS FOR: NODE 1 SHOULD MOVE UP

I understand the other argument now.
if node 3 is the one to move, it means that: when node 3 is folded and the user unfolds it, it will "jump" down.
It means that a node that you interact with can "jump" somewhere based on his children, and I can understand that it can be a bit annoying.

by the way: is this the only case that where a node "jumps" to a different location after you unfold it?

I tend to prefer @macmarrum choice, but I want to hear what others say.

**Reply by dpolivaev** [reply]
Yes, it is correct. And now you can test how it feels.
I think that in the latest preview, the jump does not hurt.

**Comment by dpolivaev** [comment]
I have found an easy solution I was not thinking about, and I am creating a preview 1.11.1_27 now so that you can test it.

**Reply by dpolivaev** [reply]
The preview is available

**Reply by macmarrum** [reply]
It looks much better :+1: 

When I look at **node 2**, I can see that it's FIRST_**CHILD-BRANCH**_BY_PARENT rather than FIRST_**CHILD-NODE**_BY_PARENT

![bug-v11-ChildNodesLayout-position-not-respected-FIRST_BRANCH_BY_PARENT](https://user-images.githubusercontent.com/7197131/229378849-33dd9d59-285e-498a-97f0-94d1ac19e492.png)

I'd very much prefer FIRST_**CHILD-NODE**_BY_PARENT, i.e. the following placement of **node 2**

![bug-v11-ChildNodesLayout-position-not-respected-FIRST_CHILD_BY_PARENT-instead-of-BRANCH](https://user-images.githubusercontent.com/7197131/229379089-1c67cf7b-82be-443f-931d-575a495a9608.png)

**Reply by dpolivaev** [reply]
All right, makes sense to me.

**Reply by dpolivaev** [reply]
Preview 1.11.1_28

**Reply by macmarrum** [reply]
Excellent :rocket:
FIRST_CHILD_BY_PARENT works great :+1:

**Comment by sifran-github** [comment]
posted a general bug which is related: https://github.com/freeplane/freeplane/issues/1075
